### DISCUSSION TITLE
UA, different recognition between online demo and demo examples in git repo

### DISCUSSION THREAD

**Discussion by KVInventoR** [ria-com/nomeroff-net#269] [General]
Різні результати розпізнавання номерів між online demo і прикладами в examples.

У мене є кілька прикладів UA номерів, я замінив зображення в get-started-demo.py на свою, 
та перевірив поведінку, отримав різні результати між online системою та локальною версією. 
Як можна покращити розпізнавання для локальної версії? 
Чи відрізняються моделі розпізнавання між online і тими що знаходяться в репозиторії?

Як можливо покращити розпізнавання для локальної версії?
Також бачу, що локальна версія не розпізнає квадратні номери, які містять 2 рядки значень.

**Comment by ApelSYN** [comment]
> Різні результати розпізнавання номерів між online demo і прикладами в examples.

Це залежить виключно від налаштувань модуля, ми використовуємо ту саму версію, що є у репозиторію у відкритому доступі. Більша функціональність досягається бульшою тратою ресурсів, тож в демо-прикладах ми взяли деякий компроміс між можливостями модуля та ресурсами/швидкісттю виконання.



> У мене є кілька прикладів UA номерів, я замінив зображення в get-started-demo.py на свою,
> та перевірив поведінку, отримав різні результати між online системою та локальною версією.

Надайте ці приклади і ми зможемо Вам відповісти як одержати бажаний результат у себе.

> Як можна покращити розпізнавання для локальної версії?

Налаштуваннями модуля.

> Чи відрізняються моделі розпізнавання між online і тими що знаходяться в репозиторії?

Є невеличка різниця, в тому що ми детектимо крім звичайних номерів ще замальовані номера, так-звані, "брендові номера" та рамки без номера. Але це, скоріш за все, впливає в гіршу сторону на якість розпізнавання.

> Також бачу, що локальна версія не розпізнає квадратні номери, які містять 2 рядки значень.

Є приклад де "квадратні" номера розпізнає, прото від більш затратний по ресурсам:
https://github.com/ria-com/nomeroff-net/blob/master/examples/ju/inference/number-plate-recognition-multiline-demo.ipynb

Головною відмінністтю наших налаштувань є використання більш складної моделі з бекбоном на архітектурі efficientnet_b2 для детекції номеру, вона працює з розміром 100x400 (по замовчуванню у  resnet18 50x200) 

Ось тут список посилань на всі відомі NomeroffNet моделі:
https://github.com/ria-com/nomeroff-net/blob/master/nomeroff_net/tools/mcm.py

Додаткові налаштування передаються у параметрі "presets":
```python
number_plate_detection_and_reading = pipeline("number_plate_detection_and_reading", image_loader="opencv", 
                 presets={
                        "eu_ua_2004_2015_efficientnet_b2": {
                            "for_regions": ["eu_ua_2015", "eu_ua_2004"],
                            "model_path": "latest"
                         }
                 }, one_preprocess_for_ocr_and_classification=False)
```

**Reply by KVInventoR** [reply]
Дякую, ця нова модель працює реально круто і дуже гарні показники розпізнавання.
Там кинув листа на пошту, якщо буде можливість, допоможіть, будь ласка, дякую.

**Comment by Ridel89** [comment]
Привіт!
Дякую вам за чудову систему. В мене та сама проблема, ніяк не можу добитися такої ж якості розпізнавання як в online demo. Чи не могли б ви скинути які саме параметри ви використовуєте в online demo?

**Comment by ApelSYN** [comment]
> Привіт! Дякую вам за чудову систему. В мене та сама проблема, ніяк не можу добитися такої ж якості розпізнавання як в online demo. Чи не могли б ви скинути які саме параметри ви використовуєте в online demo?

Ми використовуємо розподілену систему, окремі функції якої розміщені на різних серверах, тому наша конфігурація відрізняється від тієї що йде репозиторії в плані архітектури софта. Проте моделі ми використовуємо ті самі що й у відкритому доступі (за виключенням допрацьованої моделі EU, вона використовуэться у комерційному сервісі [Nomeroff Net Professional](https://ai.ria.com/ANPR-NomeroffNetProfessional) і може відрізнятись від тієї що у відкритому доступі). 

Спочатку Nomeroff Net використовував "швидкі" моделі, які менш точні проте споживають значно менше ресурсів. Це певний компроміс між швидкістю, обчислювальними ресурсами і якістю. Самі об'ємні і точні моделі мають суфікс "efficientnet_b2", моделі з суфіксом "eu_shufflenet_v2_x2_0" та без суфіксу більш прості і більш швидкі і мають менший об'єм. Останнім часом ми тренуємо нові моделі виключно з архітектурою "efficientnet_b2", і по-трохи переводимо на неї все що є.

Перелік конфігурацій всіх моделей можна переглянути ось тут:
https://github.com/ria-com/nomeroff-net/blob/master/nomeroff_net/tools/mcm.py

Я вже писав раніше, що конфігурація моделей, задається у параметрі **presets**. Є параметри по-замовчуванню, які прописані ось тут:
https://github.com/ria-com/nomeroff-net/blob/916556a7e4c397e9860910208ad375fbbca30af0/nomeroff_net/pipelines/number_plate_text_reading.py
у парметрі **DEFAULT_PRESETS**.

Підправте цей конфіг під себе:
```
presets = {
    "eu_ua_2004_2015_efficientnet_b2": {
        "for_regions": ["eu_ua_2015", "eu_ua_2004"],
        "model_path": "latest"
    },
    "eu_ua_1995_efficientnet_b2": {
        "for_regions": ["eu_ua_1995"],
        "model_path": "latest"
    },
    "eu_ua_custom_efficientnet_b2": {
        "for_regions": ["eu_ua_custom"],
        "model_path": "latest"
    },
    "xx_transit_efficientnet_b2": {
        "for_regions": ["xx_transit"],
        "model_path": "latest"
    },
    "eu_efficientnet_b2": {
        "for_regions": ["eu", "xx_unknown"],
        "model_path": "latest"
    },
    "ru": {
        "for_regions": ["ru", "eu_ua_ordlo_lpr", "eu_ua_ordlo_dpr"],
        "model_path": "latest"
    },
    "kz": {
        "for_regions": ["kz"],
        "model_path": "latest"
    },
    "kg": {  # "kg_shufflenet_v2_x2_0"
        "for_regions": ["kg"],
        "model_path": "latest"
    },
    "ge": {
        "for_regions": ["ge"],
        "model_path": "latest"
    },
    "su_efficientnet_b2": {
        "for_regions": ["su"],
        "model_path": "latest"
    },
    "am": {
        "for_regions": ["am"],
        "model_path": "latest"
    },
    "by": {
        "for_regions": ["by"],
        "model_path": "latest"
    },
}
```

Можливо у вас не зустрічаються номера am, ge, ... , можете їх прибрати з конфігу і якщо вони визначаться то їх можна опрацювати моделлю eu,  тоді конфіг виглядатиме так:
```
presets = {
    "eu_ua_2004_2015_efficientnet_b2": {
        "for_regions": ["eu_ua_2015", "eu_ua_2004"],
        "model_path": "latest"
    },
    "eu_ua_1995_efficientnet_b2": {
        "for_regions": ["eu_ua_1995"],
        "model_path": "latest"
    },
    "eu_ua_custom_efficientnet_b2": {
        "for_regions": ["eu_ua_custom"],
        "model_path": "latest"
    },
    "xx_transit_efficientnet_b2": {
        "for_regions": ["xx_transit"],
        "model_path": "latest"
    },
    "eu_efficientnet_b2": {
        "for_regions": ["eu", "xx_unknown", "am", "ge"],
        "model_path": "latest"
    },
    "ru": {
        "for_regions": ["ru", "eu_ua_ordlo_lpr", "eu_ua_ordlo_dpr"],
        "model_path": "latest"
    },
    "kz": {
        "for_regions": ["kz"],
        "model_path": "latest"
    },
    "kg": {  # "kg_shufflenet_v2_x2_0"
        "for_regions": ["kg"],
        "model_path": "latest"
    },
    "su_efficientnet_b2": {
        "for_regions": ["su"],
        "model_path": "latest"
    },
    "by": {
        "for_regions": ["by"],
        "model_path": "latest"
    },
}
```

**Reply by Ridel89** [reply]
Дякую за відповідь! Я так і використовую пресети, брав їх з number_plate_text_reading.py. Але після оновлення з версії 3.4.1 до 3.5.0 і додавання нових пресетів: eu_ua_custom_efficientnet_b2 та xx_transit_efficientnet_b2, стало гірше розпізнавати багаторядкові номера. Я використовую task - multiline_number_plate_detection_and_reading. У версії 3.4.1 багаторядкові номера розпізнавало дуже добре, а в цій чомусь видає помилку зазвичай в 2-му символі укр. номера. Пробував і без нових пресетів, результат був таким же. З чим це може бути пов'язано?

**Comment by ApelSYN** [comment]
Ми міняли реалізацію підкючення  модуля CRAFT, бо в старій реалізації є
memory leak, може щось не врахували, там пепедається багато налаштувань.
Глянемо, пришліть приклади ті що працюють на старій версії але не працюють
на новій.

пт, 31 трав. 2024 р., 15:49 користувач Ridel89 ***@***.***>
пише:
